<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="robot_name"/>

	<include file="$(find gazebo_ros)/launch/empty_world.launch">
		    <arg name="world_name" value="test1.world"/>
    		<arg name="debug" value="false" />
    		<arg name="gui" value="true" />
    		<arg name="paused" value="false"/>
    		<arg name="use_sim_time" value="true"/>
    		<arg name="headless" value="false"/>
	</include>

  <group ns="$(arg robot_name)">
    <arg name="model" default="$(find robot_description)/urdf/4omni_sim.urdf"/>   
	  <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg model)"/>

	  <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param robot_description -urdf -model $(arg robot_name) -x 0 -y 0 -z 1" />

    <!-- Load joint controller configurations from YAML file to parameter server -->
    <rosparam file="$(find robot_control)/config/4omni_control.yaml" command="load"/>

    <!-- load the controllers -->
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
      output="screen" args="wheel0_motor_shaft_joint_controller wheel1_motor_shaft_joint_controller wheel2_motor_shaft_joint_controller wheel3_motor_shaft_joint_controller"/>

    <!-- convert x-y velocity to each motor velocity -->
    <node name="trans" pkg="robot_control" type="4omni_trans.py">
        <remap from="/wheel0" to="wheel0_motor_shaft_joint_controller/command" />
        <remap from="/wheel1" to="wheel1_motor_shaft_joint_controller/command" />
        <remap from="/wheel2" to="wheel2_motor_shaft_joint_controller/command" />
        <remap from="/wheel3" to="wheel3_motor_shaft_joint_controller/command" />
        <param name="radius" value="0.1"/>
        <param name="center_to_wheel" value="0.25"/>
    </node>

    <!-- publish odom -->
    <node name="gazebo_odom" pkg="robot_control" type="gazebo_odom"  output="screen">
      <param name="model_name"  value="$(arg robot_name)" />
      <param name="odom_frame" value="$(arg robot_name)/odom" />
      <param name="base_frame"  value="$(arg robot_name)/base_link" />
      <param name="publish_rate"  value="200" />
      <param name="noise"  value="0.0" />
      <param name="tf_enable" value="true" />
      <remap from="odom" to="odom" />
    </node>

  </group>
</launch>
